core/variables_dynamic.js

```diff
@@ -0,0 +1,127 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Utility functions for handling variables dynamic.
+ *
+ * @author duzc2dtw@gmail.com (Du Tian Wei)
+ */
+'use strict';
+
+goog.provide('Blockly.VariablesDynamic');
+
+goog.require('Blockly.Variables');
+goog.require('Blockly.Blocks');
+goog.require('Blockly.constants');
+goog.require('Blockly.VariableModel');
+goog.require('Blockly.Workspace');
+goog.require('goog.string');
+
+
+/**
+ * Prompt the user for a new variable name.
+ * @param {string} promptText The string of the prompt.
+ * @param {string} defaultText The default value to show in the prompt's field.
+ * @param {function(?string)} callback A callback. It will return the new
+ *     variable name, or null if the user picked something illegal.
+ */
+Blockly.VariablesDynamic.promptType = function(promptText, defaultText, callback) {
+  Blockly.prompt(promptText, defaultText, function(newVarType) {
+    // Merge runs of whitespace.  Strip leading and trailing whitespace.
+    // Beyond this, all types are legal.
+    if (newVarType) {
+      newVarType = newVarType.replace(/[\s\xa0]+/g, ' ').replace(/^ | $/g, '');
+    }
+    callback(newVarType);
+  });
+};
+Blockly.VariablesDynamic.onCreateVariableButtonClick = function(button) {
+  Blockly.VariablesDynamic.promptType(Blockly.Msg.NEW_VARIABLE_TYPE_TITLE, '', function(type) {
+    if (type) {
+      Blockly.Variables.createVariable(button.getTargetWorkspace(), null, type);
+    }
+  });
+  // workspace.createVariable("abc", "string");
+  // workspace.createVariable("123", "number");
+  // workspace.createVariable("abcd", "string");
+};
+/**
+ * Construct the elements (blocks and button) required by the flyout for the
+ * variable category.
+ * @param {!Blockly.Workspace} workspace The workspace contianing variables.
+ * @return {!Array.<!Element>} Array of XML elements.
+ */
+Blockly.VariablesDynamic.flyoutCategory = function(workspace) {
+  var xmlList = [];
+  var button = goog.dom.createDom('button');
+  button.setAttribute('text', Blockly.Msg.NEW_VARIABLE);
+  button.setAttribute('callbackKey', 'CREATE_VARIABLE');
+
+  workspace.registerButtonCallback('CREATE_VARIABLE', Blockly.VariablesDynamic.onCreateVariableButtonClick);
+
+  xmlList.push(button);
+
+  var blockList = Blockly.VariablesDynamic.flyoutCategoryBlocks(workspace);
+  xmlList = xmlList.concat(blockList);
+  return xmlList;
+};
+
+/**
+ * Construct the blocks required by the flyout for the variable category.
+ * @param {!Blockly.Workspace} workspace The workspace contianing variables.
+ * @return {!Array.<!Element>} Array of XML block elements.
+ */
+Blockly.VariablesDynamic.flyoutCategoryBlocks = function(workspace) {
```

Comment:
Since any setter can be used for any variable and the type isn't indicated in any way did you find it useful having a setter per type? Could Blockly.Variables.flyoutCategoryBlocks() be reused instead?

Reply:
Sorry , I don't think so .
Blockly.Variables.flyoutCategoryBlocks should display only  untyped or '' typed variables.
There are very different behaviour when user make a typed variable.
when I drag a typed setter block,  I konw what type of variables I should use , and the dropdown menu should not mix my mind . and that makes the dropdown menu clear when I have too much variables .

Reply:
Whoops, I failed to say what I meant. I think this should use the same logic as Blockly.Variables.flyoutCategoryBlocks, just using 'variables_set_dynamic' and 'variables_get_dynamic' instead. I don't see any benefit to listing a setter for each type if there's no way for the user to know why they have those specific setters.

Example:
If I have variables with types:
foo: int
bar: int
baz: string
loo: string
lar: string
lan: boolean

Then it would show the following blocks:
set foo
set baz
set lan
foo
bar
baz
loo
lar
lan

From just that it's not clear why foo, baz, and lan have setters and the rest don't. I'd prefer a single setter like the regular variables category for now and we explore ways to identify types in a later PR.

Reply:
Sorry again for the confusion around this. Once you add the fix to field_variable.js only having a single setter should be viable.